src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    Node<T> node(int index) {
+        Node<T> find;
+        if (index < (size / 2)) {
+            find = head;
+            for (int i = 0; i < index; i++) {
+                find = find.next;
+            }
+        } else {
+            find = tail;
+            for (int i = size - 1; i > index; i--) {
+                find = find.prev;
+            }
+        }
+        return find;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> t = tail;
+        Node<T> newNode = new Node<>(t, value, null);
+        tail = newNode;
+        if (t == null) {
+            head = newNode;
+        } else {
+            t.next = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
 
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> actual = node(index);
+            Node<T> actualPrev = actual.prev;
+            Node<T> newNode = new Node<>(actualPrev, value, actual);
+            node(index).prev = newNode;
+            if (actualPrev == null) {
+                head = newNode;
+            } else {
+                actualPrev.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkPositionIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkPositionIndex(index);
+        Node<T> x = node(index);
+        T oldVal = x.item;
+        x.item = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkPositionIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> x = head; x != null; x = x.next) {
+            if (object == x.item || x.item.equals(object)) {
```

Comment:
can we get NPE here?

Reply:
in the case when we use condition " x != null "  we talk that our node can't be null, but the item in the node can be null. `x.item` can return null, so when we call `x.item.equals(object)` we will get NPE